### PR TITLE
feat(ci): add `sending_transactions_using_lightwalletd` test to CI

### LINKED ISSUE #3926: ci: run a full lightwalletd sync from a close-to-tip zebrad
## Motivation
On every PR push, if the zebrad state format hasn’t changed, we want to run a full lightwalletd sync from a close-to-tip zebrad, as a test. This test should take 30 minutes

### Designs

- Lightwalletd updates from zebrad in less than half an hour, so that part of the test can be  run on every PR change
- This approach would use already generated disks from previous stateful disk generations job https://github.com/ZcashFoundation/zebra/blob/main/.github/workflows/test.yml#L213
- A `lightwalletd-test.yml` might be appropriate from this point on and move specific lightwalletd tests here.

## Related Work
https://github.com/ZcashFoundation/zebra/issues/3922

### LINKED ISSUE #4168: ci: call lightwalletd wallet queries
## Motivation

We need to call the `lightwalletd_wallet_grpc_tests` tests from #3655 in CI, using the lightwalletd cached state from #4165.

(PR #4267 already added the tests from #3512.)

### LINKED ISSUE #4305: Set RUST_LOG environmental variable in gcp-test-deploy.yml
## Motivation

We can't see when GCP tests are being skipped, because the test logs aren't active in gcp-test-deploy.yml

### Designs

Use the same logging environmental variables as the build workflows:
```
RUST_BACKTRACE="full"
RUST_LIB_BACKTRACE="full"
COLORBT_SHOW_HIDDEN="1"
ZEBRA_SKIP_IPV6_TESTS="1"
RUST_LOG="info"
```

We should hard-code these values for now.

### PR DESCRIPTION
**Pull request by gustavovalverde**
## Motivation

We want to add all completed `lightwalletd` integration tests to CI, to confirm we're not introducing bugs to the RPC endpoints or any other implementation related to `lightwalletd`

Fixes #4168
Fixes: #4305

### Designs

- Consider the implementations being done in `test.yml` and `test-full-sync.yml`

## Solution

- Add `sending_transactions_using_lightwalletd` test to CI
- Add `sending_transactions_using_lightwalletd` to `entrypoint.sh` for easier execution

## Review

Anyone from the DevOps team can review this

### Reviewer Checklist

  - [ ] Verify tests are using the right execution methods
  - [ ] Confirm job id, step names and description

## Follow Up Work

Simplify this job with the refactor being done for reusable workflows

### CODE REVIEW

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -215,3 +215,18 @@ jobs:
...
+  # Test sending transactions with lightwalletd using a Zebra tip state
+  lightwalletd-transactions-test:
+    if: ${{ github.event.inputs.regenerate-disks != 'true' }}
+    needs: build
+    uses: ./.github/workflows/gcp-test-deploy.yml
+    with:
+      app_name: lightwalletd
+      test_id: lwd-send-transactions
+      test_description: Test lightwalletd transactions with a Zebra tip state
+      test_variables: '-e TEST_LWD_TRANSACTIONS=1 -e ZEBRA_TEST_LIGHTWALLETD=1 -e ZEBRA_FORCE_USE_COLOR=1 -e ZEBRA_CACHED_STATE_DIR=/var/cache/zebrad-cache'
```
Comment (teor2345) [inline .github/workflows/test.yml]: This test needs a lightwalletd directory path, but that directory does not need to exist:

```suggestion
      test_variables: '-e TEST_LWD_TRANSACTIONS=1 -e ZEBRA_TEST_LIGHTWALLETD=1 -e ZEBRA_FORCE_USE_COLOR=1 -e ZEBRA_CACHED_STATE_DIR=/var/cache/zebrad-cache -e LIGHTWALLETD_DATA_DIR=/var/cache/lightwalletd-cache'
```

**Review by teor2345 — DISMISSED** [review]
We're missing an env var, then we should be good.

Does this workflow have the same state connection bug as PR #4223?

File: .github/workflows/test.yml [hunk 2]
```diff
@@ -217,7 +217,6 @@ jobs:
       height_grep_text: 'finished initial sync to chain tip, using gossiped blocks sync_percent=100.* current_height=Height'
 
   lightwalletd-rpc-test:
-    if: ${{ github.event.inputs.regenerate-disks != 'true' }}
```
Comment (teor2345) [inline .github/workflows/test.yml:232]: Question: why is this condition removed?

File: .github/workflows/test.yml [hunk 3]
```diff
@@ -228,4 +227,20 @@ jobs:
...
       saves_to_disk: false
       disk_suffix: tip
-      zebra_state_path: '/var/cache'
+      root_state_path: '/var/cache'
+      zebra_state_dir: 'zebrad-cache'
+
+  lightwalletd-transactions-test:
+    needs: build
+    uses: ./.github/workflows/gcp-test-deploy.yml
+    if: ${{ github.event.inputs.regenerate-disks != 'true' || github.event.inputs.run-full-sync != 'true' }}
```
Comment (teor2345) [inline .github/workflows/test.yml]: Question: why is this condition different from all the other jobs?

(It's ok to make them all the same in this PR.)

File: .github/workflows/test.yml [hunk 4]
```diff
@@ -228,4 +227,20 @@ jobs:
...
+    with:
+      app_name: lightwalletd
+      test_id: lwd-send-transactions
+      test_description: Test lightwalletd transactions with a Zebra tip state
```
Comment (teor2345) [inline .github/workflows/test.yml]:
```suggestion
      test_description: Test sending transactions via lightwalletd
```

File: .github/workflows/test.yml [hunk 5]
```diff
@@ -217,7 +217,6 @@ jobs:
       height_grep_text: 'finished initial sync to chain tip, using gossiped blocks sync_percent=100.* current_height=Height'
 
   lightwalletd-rpc-test:
```
Comment (teor2345) [inline .github/workflows/test.yml:232]:
```suggestion
  # Test that Zebra can answer a synthetic RPC call, using a cached Zebra tip state
  lightwalletd-rpc-test:
```

File: .github/workflows/test.yml [hunk 6]
```diff
@@ -228,4 +227,20 @@ jobs:
       needs_zebra_state: true
       saves_to_disk: false
       disk_suffix: tip
-      zebra_state_path: '/var/cache'
+      root_state_path: '/var/cache'
+      zebra_state_dir: 'zebrad-cache'
+
+  lightwalletd-transactions-test:
```
Comment (teor2345) [inline .github/workflows/test.yml:249]:
```suggestion
  # Test that Zebra can handle a lightwalletd send transaction RPC call, using a cached Zebra tip state
  # TODO: use a cached lightwalletd tip state to speed up the test (#4303)
  lightwalletd-transactions-test:
```

File: .github/workflows/zcash-lightwalletd.yml [hunk 7]
```diff
@@ -110,4 +110,4 @@ jobs:
           labels: ${{ steps.meta.outputs.labels }}
           push: true
           cache-from: type=registry,ref=${{ env.GAR_BASE }}/${{ env.IMAGE_NAME }}:buildcache
-          cache-to: type=registry,ref=${{ env.GAR_BASE }}/${{ env.IMAGE_NAME }}:buildcache,mode=max
+          cache-to: type=registry,ref=${{ env.GAR_BASE }}/${{ env.IMAGE_NAME }}:buildcache,mode=max
```
Comment (teor2345) [inline .github/workflows/zcash-lightwalletd.yml]: Question: Why is this change needed? Do we need to open a ticket to fix a workflow dependency?

File: zebrad/tests/acceptance.rs [hunk 8]
```diff
@@ -1526,6 +1526,7 @@ async fn fully_synced_rpc_test() -> Result<()> {
 ///
 /// See [`common::lightwalletd::send_transaction_test`] for more information.
 #[tokio::test]
+#[ignore]
```
Comment (teor2345) [inline zebrad/tests/acceptance.rs:1529]: This is good for consistency, we might want to do it for `lightwalletd_wallet_grpc_tests` and `lightwalletd_update_sync` as well.

**Review by teor2345 — COMMENTED** [review]
This is looking good so far, I just have a few questions.

Edit: question resolved by PR #4332

**Comment by teor2345** [comment]
@jvff @gustavovalverde it seems like there is a bug in the `sending_transactions_using_lightwalletd` test:
> State directory doesn't have a chain tip block

This state path looks suspicious:
> May 05 20:03:11.748  INFO Opened Zebra state cache at /tmp/zebrad_testsDJTecY/state/state/v23/mainnet

- `state/state` is duplicated, Zebra only uses one `state` directory by default
- `/tmp/zebrad_testsDJTecY` is a temporary directory, should it be using `/var/cache/zebrad-cache` instead?

It would be helpful if `sending_transactions_using_lightwalletd` logged what it was doing at info level.

https://github.com/ZcashFoundation/zebra/runs/6312788960?check_suite_focus=true#step:8:217

**Comment by teor2345** [comment]
> Why is the full-sync-to-tip test running in this PR's GitHub actions?
> Did we make a mistake in the run conditions?
> https://github.com/ZcashFoundation/zebra/runs/6312789433?check_suite_focus=true

I opened PR #4332 to fix this.

File: .github/workflows/test.yml (hunk 2 above)
Comment (gustavovalverde) [reply .github/workflows/test.yml:232]: It was wrongly place (and wrongly removed). Fixed now.

File: .github/workflows/test.yml [hunk 9]
```diff
@@ -216,10 +216,11 @@ jobs:
       disk_suffix: tip
       height_grep_text: 'finished initial sync to chain tip, using gossiped blocks sync_percent=100.* current_height=Height'
 
+  # Test that Zebra can answer a synthetic RPC call, using a cached Zebra tip state
   lightwalletd-rpc-test:
-    if: ${{ github.event.inputs.regenerate-disks != 'true' }}
     needs: build
     uses: ./.github/workflows/gcp-test-deploy.yml
+    if: ${{ github.event.inputs.regenerate-disks != 'true' }}
```
Comment (teor2345) [inline .github/workflows/test.yml]:
```suggestion
    if: ${{ github.event.inputs.regenerate-disks != 'true' && github.event.inputs.run-full-sync != 'true' }}
```

File: .github/workflows/test.yml [hunk 10]
```diff
@@ -228,4 +229,22 @@ jobs:
...
-      zebra_state_path: '/var/cache'
+      root_state_path: '/var/cache'
+      zebra_state_dir: 'zebrad-cache'
+
+  # Test that Zebra can handle a lightwalletd send transaction RPC call, using a cached Zebra tip state
+  # TODO: use a cached lightwalletd tip state to speed up the test (#4303)
+  lightwalletd-transactions-test:
+    needs: build
+    uses: ./.github/workflows/gcp-test-deploy.yml
+    if: ${{ github.event.inputs.regenerate-disks != 'true' }}
```
Comment (teor2345) [inline .github/workflows/test.yml]:
```suggestion
    if: ${{ github.event.inputs.regenerate-disks != 'true' && github.event.inputs.run-full-sync != 'true' }}
```

**Review by teor2345 — COMMENTED** [review]
I think these are the conditions we want

File: .github/workflows/test.yml [hunk 11]
```diff
@@ -190,7 +190,7 @@ jobs:
   test-stateful-sync:
     needs: build
     uses: ./.github/workflows/gcp-test-deploy.yml
-    if: ${{ github.event.inputs.regenerate-disks != 'true' }}
+    if: ${{ github.event.inputs.regenerate-disks != 'true' || github.event.inputs.run-full-sync != 'true' }}
```
Comment (teor2345) [inline .github/workflows/test.yml]: We only want to run these tests when we are not running any manual tests

```suggestion
    if: ${{ github.event.inputs.regenerate-disks != 'true' && github.event.inputs.run-full-sync != 'true' }}
```

**Comment by teor2345** [comment]
>  - Quota 'C2D_CPUS' exceeded.  Limit: 300.0 in region us-central1.

**Comment by teor2345** [comment]
> > * Quota 'C2D_CPUS' exceeded.  Limit: 300.0 in region us-central1.

I'm re-running those jobs now.

**Comment by teor2345** [comment]
Still failing with:
> May 05 22:28:14.398  INFO Opened Zebra state cache at /tmp/zebrad_testseZOF6F/state/state/v23/mainnet
> ...
> State directory doesn't have a chain tip block

https://github.com/ZcashFoundation/zebra/runs/6314502890?check_suite_focus=true#step:8:292

**Comment by teor2345** [comment]
@mergifyio update

File: zebrad/tests/common/lightwalletd/wallet_grpc.rs [hunk 12]
```diff
@@ -42,7 +42,7 @@ pub fn spawn_lightwalletd_with_rpc_server(
 
     lightwalletd.expect_stdout_line_matches("Starting gRPC server")?;
     if wait_for_blocks {
```
Comment (teor2345) [inline zebrad/tests/common/lightwalletd/wallet_grpc.rs:44]: This condition might not be needed any more, but we can check that when we get to the other grpc tests.

**Comment by teor2345** [comment]
> Error: buildx failed with: error: failed to solve: failed to fetch oauth token: unexpected status: 401 Unauthorized

I'll restart

**Comment by teor2345** [comment]
@mergifyio update

**Review by teor2345 — APPROVED** [review]
This all appears to be working correctly

> {"app":"lightwalletd","level":"info","msg":"Adding new mempool txid 0f0e4c4978d9446bc9b173612c71c705ecf7dd2f6e7432f6b04697d6b2be9a04  sending to  0  clients","time":"2022-05-06T04:13:04Z"}

https://github.com/ZcashFoundation/zebra/runs/6317043136?check_suite_focus=true#step:8:728

> test sending_transactions_using_lightwalletd ... ok
> test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 37 filtered out; finished in 2241.01s

https://github.com/ZcashFoundation/zebra/runs/6317043136?check_suite_focus=true#step:8:819

> May 06 04:13:04.654  INFO {net="Main"}: zebrad::components::mempool::gossip: sending mempool transaction broadcast request=AdvertiseTransactionIds({Legacy(transaction::Hash("0f0e4c4978d9446bc9b173612c71c705ecf7dd2f6e7432f6b04697d6b2be9a04"))})

https://github.com/ZcashFoundation/zebra/runs/6317043136?check_suite_focus=true#step:8:812
